### PR TITLE
Spotinstances

### PR DESCRIPTION
**Pull request by icarthick**
Issue #, if available:
https://github.com/awslabs/aws-simple-ec2-cli/issues/55

Description of changes:
This PR includes the ability to request for Spot Instances. To accomplish this, I have introduced a new command line arg called "Ec2 Purchase Instance Type". This is an enum which supports one of these two choices "OnDemand" and "SpotInstance". If "SpotInstance" is chosen as the EC2 Purchase instance type, only then it would prompt the user for Spot Price. Depending on the spot price provided by the user, the Spot Instance Requests may or may not succeed. By default OnDemand instances will be chosen unless otherwise specified.

We are running out of flags, I have chosen "z" as the short flag name to specify Ec2 Purchase Instance Type and "x" as the short flag name for specifying Spot Instance Price. Am happy to change them if needed

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by snay2 — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a good start.

I'm not sold yet on whether this is the right "simple-ec2"-flavored customer experience for launching Spot--we ask the user for a Spot bid price separately from the instance type, without guidance on what those choices should be, and we're not much help if that Spot request fails (should they choose a different instance type? Increase their bid? Use a different region/AZ? Give up and use On Demand?). I wonder if we can provide more guidance or do some additional heavy lifting for the user, so that even someone who's never used Spot before can run it successfully with the defaults.

Spot is a potentially complex product, so I want to make sure we can provide a simple enough experience while still allowing the necessary flexibility.

File: cmd/launch.go [hunk 1]
```diff
@@ -76,7 +83,14 @@ func launch(cmd *cobra.Command, args []string) {
 	}
 
 	// Start a new session, with the default credentials and config loading
-	sess := session.Must(session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable}))
+	creds := processcreds.NewCredentials("isengardcli credentials --awscli --role Admin carthick+webeip1@amazon.com")
```
[the remark this answers is not in the document]
Comment (icarthick) [reply cmd/launch.go]: Good catch, pushed the change

File: README.md [hunk 2]
```diff
@@ -105,12 +105,18 @@ Flags:
   -g, --security-group-ids strings       The security groups with which the instance will be launched
   -s, --subnet-id string                 The subnet id in which the instance will be launched
       --tags stringToString              The tags applied to instances and volumes at launch (Example: tag1=val1,tag2=val2) (default [])
+  -x, --spot-price float                 The price at which spot instances
```
[the remark this answers is not in the document]
Comment (icarthick) [reply README.md]: Addressed

File: pkg/cli/cli.go [hunk 3]
```diff
@@ -27,6 +27,8 @@ const (
 
 // Enum values for displaying resource types in CLI
 const (
+	Ec2PurchanceInstanceType         = "Ec2 Purchase Instance Type"
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/cli/cli.go]: Addressed

File: pkg/config/config.go [hunk 4]
```diff
@@ -34,6 +35,8 @@ A simple config for reading config files or flags into primitive type informatio
 The config will later be used to parse into a detailed config and to launch an instance.
 */
 type SimpleInfo struct {
+	Ec2PurchaseInstanceType       Ec2PurchseInstanceType
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/config/config.go]: Addressed

File: pkg/ec2helper/ec2helper.go [hunk 5]
```diff
@@ -1220,26 +1311,26 @@ func (h *EC2Helper) createNetworkConfiguration(simpleConfig *config.SimpleInfo,
...
+		return nil, nil, errors.New("Unknown security group placeholder")
 	}
 
 	if len(selectedSecurityGroupIds) <= 0 {
-		return errors.New("No security group available for stack")
+		return nil, nil, errors.New("No security group available for stack")
 	}
 
-	input.SecurityGroupIds = aws.StringSlice(selectedSecurityGroupIds)
+	//input.SecurityGroupIds = aws.StringSlice(selectedSecurityGroupIds)
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/ec2helper/ec2helper.go]: Addressed

File: cmd/launch.go [hunk 6]
```diff
@@ -44,6 +47,10 @@ func init() {
 	launchCmd.Flags().BoolVarP(&isInteractive, "interactive", "i", false, "Interactive mode")
 	launchCmd.Flags().StringVarP(&flagConfig.Region, "region", "r", "",
 		"The region where the instance will be launched")
+	launchCmd.Flags().VarP(&flagConfig.Ec2PurchaseInstanceType, "purchase-instance-type","z",
+		`The purchase type of the instance. Allowed values "On Demand" and "Spot Instance"`)
```
[the remark this answers is not in the document]
Comment (icarthick) [reply cmd/launch.go:49]: Addressed

File: cmd/launch.go [hunk 7]
```diff
@@ -87,13 +101,42 @@ func launch(cmd *cobra.Command, args []string) {
...
+
+
 // Launch the instance interactively
 func launchInteractive(h *ec2helper.EC2Helper) {
 	var err error
 	simpleConfig := config.NewSimpleInfo()
 	// Override config with flags if applicable
 	config.OverrideConfigWithFlags(simpleConfig, flagConfig)
 
+	if simpleConfig.Ec2PurchaseInstanceType == "" {
```
[the remark this answers is not in the document]
Comment (icarthick) [reply cmd/launch.go]: Addressed

File: pkg/ec2helper/ec2helper.go [hunk 8]
```diff
@@ -987,6 +988,52 @@ func (h *EC2Helper) ParseConfig(simpleConfig *config.SimpleInfo) (*config.Detail
...
+		if HasEbsVolume(detailedConfig.Image) && simpleConfig.KeepEbsVolumeAfterTermination {
+			input.LaunchSpecification.BlockDeviceMappings = detailedConfig.Image.BlockDeviceMappings
+			for _, block := range input.LaunchSpecification.BlockDeviceMappings {
+				if block.Ebs != nil {
+					block.Ebs.DeleteOnTermination = aws.Bool(false)
+				}
+			}
+		}
+	}
+	if simpleConfig.BootScriptFilePath != "" {
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/ec2helper/ec2helper.go]: Addressed

File: pkg/config/config.go [hunk 9]
```diff
@@ -104,6 +141,17 @@ func OverrideConfigWithFlags(simpleConfig *SimpleInfo, flagConfig *SimpleInfo) {
...
+	if flagConfig.Ec2PurchaseInstanceType != "" {
+		simpleConfig.Ec2PurchaseInstanceType = flagConfig.Ec2PurchaseInstanceType
+	}
+	if flagConfig.SpotInstancePrice != 0 {
+		simpleConfig.SpotInstancePrice = flagConfig.SpotInstancePrice
+	}
+	if flagConfig.Ec2PurchaseInstanceType == SpotInstance{
+		flagConfig.LaunchTemplateVersion = ""
+		flagConfig.LaunchTemplateId = ""
+		flagConfig.AutoTerminationTimerMinutes = 0
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/config/config.go]: Addressed

File: pkg/config/config.go [hunk 10]
```diff
@@ -49,6 +52,40 @@ type SimpleInfo struct {
 	UserTags                      map[string]string
 }
 
+type Ec2PurchseInstanceType string
+
+
+func ToPurchaseInstanceType(enumString string) Ec2PurchseInstanceType{
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/config/config.go]: Addressed

File: pkg/ec2helper/ec2helper.go [hunk 11]
```diff
@@ -1192,10 +1283,10 @@ func (h *EC2Helper) createNetworkConfiguration(simpleConfig *config.SimpleInfo,
 		}
 	}
 	if selectedSubnetId == nil {
-		return errors.New("No subnet with the selected availability zone found")
+		return nil, nil, errors.New("No subnet with the selected availability zone found")
 	}
 
-	input.SubnetId = selectedSubnetId
+	//input.SubnetId = selectedSubnetId
```
[the remark this answers is not in the document]
Comment (icarthick) [reply pkg/ec2helper/ec2helper.go:1318]: Addressed

**Comment by brycahta** [comment]
@icarthick -- Appreciate the PR and helping us build out this feature, but I share @snay2 's concerns with respect to Spot's complexity. Instead of a PR to start, the next steps for implementing Spot would be a design document/proposal outlining various details such as: which underlying api to use and why, what level of abstraction should we implement (i.e. I want to request the cheapest spot), should spot be its own subcommand etc. 

Our team plans to submit a PR containing the design proposal for the community to review in the near future, and I will update the original issue communicating the same. With that said, I'm closing this PR for now and if you have any other ideas/comments feel free to comment on the issue or upcoming design.

**Pull request closed without merging**